Imports Extension: Use the jscodeshift `use` api.

wbinnssmith opened the pull request:
**This breaks this plugin's api.** It also entirely removes the dependency on
jscodeshift itself. Instead, it exports a factory function that takes a config,
now also functions, and returns a `use`-able plugin to be registered with
jscodeshift.

As a result of moving away from the user invoking `registerMethods` directly,
we lose the ability to provide a custom name to the plugin's methods, but we
also avoid leaking the registration process to the user, and things are much
simpler IMHO.

wbinnssmith commented:
cc @pieterv @voideanvalue

cpojer commented:
Added @voideanvalue as a contributor to this project. I'll let Pieter and Bhuwan handle this PR :)

pieterv commented on extensions/imports/config/CJSBasicRequireConfig.js:
```diff
@@ -4,19 +4,20 @@ const StringUtils = require('nuclide-format-js-base/lib/utils/StringUtils');
-    ),
-  },
-];
+module.exports = function (jscs) {
```
I think just use an arrow func here, it makes it pretty nice and clean:
```
module.exports = jscs => [
  ...
];
```

wbinnssmith replied:
👍

pieterv commented on extensions/imports/config/FBRequireConfig.js:
```diff
@@ -4,34 +4,35 @@ const StringUtils = require('nuclide-format-js-base/lib/utils/StringUtils');
-      getDeclarationName(node2)
-    ),
-  },
+module.exports = function (jscs) {
```
arrow func as per the other config.

pieterv commented on extensions/imports/index.js:
```diff
@@ -4,15 +4,13 @@ const addImport = require('./addImport');
-    const plugins = this.createPlugins(config);
-    jscodeshift.registerMethods(plugins);
-  },
+function imports(makeConfig) {
```
Maybe add a quick comment here on the usage.

pieterv commented:
@wbinnssmith this is really cool! Thanks for doing this. Looks good apart from the small inline comments i made. Could you also update the readme with the new usage.

@fkling @cpojer @voideanvalue what do you think about plugin adding API here?
```
const imports = require('jscodeshift-imports');

jscodeshift.use(imports(imports.config.FBRequire))
```
Im wondering if we should move the imports plugin creation fn to a property instead of being the root since this can be kinda awkward with es6 imports. Thoughts?

wbinnssmith commented:
@pieterv heh, forgot to mention to check out the README and tests for usage.

Here it is:

```js
const imports = require('jscodeshift-imports');

module.exports = function(fileInfo, api) {
  const {jscodeshift} = api;

  jscodeshift.use(imports(imports.config.CJSBasicRequire));

  // Your transform here.
}
```

pieterv commented:
Oh actually just noticed you missed some references to jscodeshift in `utils/getDeclarationName.js` and `utils/isValidRequireDeclaration.js`, these utils are going to be a bit awkward to pass jscodeshift into but i guess we can. Also note this won't kill jscodeshift as a dependency of this plugin since @kyldvs's `format-js-base` utils are used which also require jscodeshift :/ This is what i was talking about when i said its super awkward to pass jscs to all the places that need it.

pieterv commented on extensions/imports/index.js:
```diff
@@ -4,15 +4,15 @@ const addImport = require('./addImport');
-    jscodeshift.registerMethods(plugins);
-  },
+// the main entry, which takes a config fn and returns a `use`-able plugin fn
+// ex: jscodeshift.use(imports(imports.config.CJSBasicRequire));
```
Oh, please use the `/**` style i'm using for the `addImport` API.

```
/**
 * Create extension. Takes a config fn and returns a `use`-able plugin fn.
 * 
 * Usage:
 *
 *   jscodeshift.use(imports(imports.config.CJSBasicRequire));
 */
```

pieterv commented on extensions/imports/config/FBRequireConfig.js:
```diff
@@ -4,34 +4,35 @@ const StringUtils = require('nuclide-format-js-base/lib/utils/StringUtils');
-    ),
-  },
+module.exports = jscs => {
+  return [
```
You can skip the return and the block just output the array directly:
```
module.exports = jscs => [
  ...
];
```

pieterv commented on extensions/imports/config/CJSBasicRequireConfig.js:
```diff
@@ -4,19 +4,20 @@ const StringUtils = require('nuclide-format-js-base/lib/utils/StringUtils');
-    ),
-  },
-];
+module.exports = jscs => {
```
same here too.

fkling commented on extensions/imports/package.json:
```diff
@@ -13,7 +13,9 @@
     "import"
   ],
   "dependencies": {
-    "jscodeshift": "^0.3.0",
```
I believe jscodeshift should still be a peer dependency. After all, the plugin depends on specific version of the API.

pieterv commented:
@wbinnssmith Is this code good to do a final review?

cpojer commented:
Is there any progress here? cc @wbinnssmith

wbinnssmith commented:
hey all, apologies for the delay.

Just rebased this and moved jscodeshift to a peer dependency. Looked over things and still feeling quite good (see the updated tests and README for examples).

> Also note this won't kill jscodeshift as a dependency of this plugin since @kyldvs's format-js-base utils are used which also require jscodeshift :/

Good point, though this does free us up to pass whichever jscodeshift (which may or may not be the exact object that `nuclide-format-js-base` depends on), which should prevent these kinds of things in the future.

I'll probably run this by @fkling in person soon, but I feel 👍 about a merge.
